ipatests/prci_definitions/prci_checker.py

```diff
@@ -0,0 +1,254 @@
+
+
+def print_field_error(
+    jobname, fieldname=None, expected_value=None, custom_msg=None
+):
+    """Helper function to print field errors"""
+    msg = f"In job '{jobname}':\n"
+    if custom_msg:
+        msg += f"  {custom_msg}"
+    elif fieldname and expected_value:
+        msg += (
+            f'  Job field "{fieldname}" should be defined as: '
+            f'"{fieldname}: {expected_value}"'
+        )
+    else:
+        msg = f"In job '{jobname}'."
+    print_error(msg)
+
+
+def check_jobs(filename, def_suite, current_spec):
+    """
+    Check if given job definition file has all jobs correctly defined according
+    to specification file.
+
+    :param filename: file name of the definition file to be checked
+    :param def_suite: definition file as a dict object
+    :param jobs_spec: PRCI specification file containing correct definitions
+    """
+    correct_fields = True
+
+    jobs_def = def_suite.get("jobs")
+    if jobs_def is None:
+        print_error(
+            f"'{filename}' file doesn't have a jobs section following "
+            "the format."
+        )
+        return False
+    # Print file to be analyzed and its number of jobs
+    print("[File] " + filename + " [Jobs] " + str(len(jobs_def)))
+
+    try:
+        job_prefix = current_spec["job_prefix"]
+        requires = current_spec["requires"]
+        build_url = current_spec["build_url"]
+    except KeyError as e:
+        print_error(
+            "Specification file has bad format "
+            f"and '{filename}' could not be analyzed.\n"
+            f"  KeyError: {e} in '{filename}'"
+        )
+        return False
+
+    # Get template from build job
+    build_job = job_prefix + "build"
+    build_args = jobs_def[build_job]["job"]["args"]
+    template = build_args["template"]
+
+    copr = build_args.get("copr")
+    update_packages = current_spec.get("update_packages")
+    selinux = current_spec.get("selinux_enforcing")
+
+    for job_name, params in jobs_def.items():
+        if job_name == build_job:  # Build job not analyzed
+            continue
+        args = params.get("job").get("args")
+        if not job_name.startswith(job_prefix):
+            msg = f"Job name should start with prefix '{job_prefix}'"
+            print_field_error(job_name, custom_msg=msg)
+            correct_fields = False
+        if params.get("requires") != requires:
+            print_field_error(job_name, "requires", requires)
+            correct_fields = False
+        if args.get("build_url") != build_url:
+            print_field_error(job_name, "build_url", f"'{build_url}'")
+            correct_fields = False
+        # Check template field against build target
+        if args.get("template") != template:
+            print_field_error(job_name, "template", template)
+            correct_fields = False
+        # If build target has a copr repo, check that the job also defines it
+        if copr is not None and args.get("copr") != copr:
```

Comment:
This test doesn't detect if copr is set while it should not.

Reply:
Addressed

Reply:
The checker also needs to ensure that the build job defines the copr repo for 389ds, pki and sssd nightly defs. Otherwise the checker detects inconsistencies between build and run jobs (copr defined in one but not the other), but does not detect that the copr is missing from the build job.